<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <meta name="description" content="[HIGH LEVEL DISCRIPTION.]">
        <meta name="author" content="St. Jude Cloud">
        
        <link rel="shortcut icon" href="img/favicon.ico">
        <title>[PROJECT NAME] - St. Jude</title>
        <link href="css/bootstrap.min.css" rel="stylesheet">
        <link href="css/font-awesome.min.css" rel="stylesheet">
        <link href="css/base.css" rel="stylesheet">
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/purebasic.min.css">

        <script src="js/jquery-1.10.2.min.js" defer></script>
        <script src="js/bootstrap.min.js" defer></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
        <script>hljs.initHighlightingOnLoad();</script> 
    </head>

    <body class="homepage">
        <div class="navbar fixed-top navbar-expand-lg navbar-dark bg-primary">
            <div class="container">
                <a class="navbar-brand" href=".">[PROJECT NAME] - St. Jude</a>
                <!-- Expander button -->
                <button type="button" class="navbar-toggler" data-toggle="collapse" data-target="#navbar-collapse">
                    <span class="navbar-toggler-icon"></span>
                </button>

                <!-- Expanded navigation -->
                <div id="navbar-collapse" class="navbar-collapse collapse">

                    <ul class="nav navbar-nav ml-auto">
                        <li class="nav-item">
                            <a href="#" class="nav-link" data-toggle="modal" data-target="#mkdocs_search_modal">
                                <i class="fa fa-search"></i> Search
                            </a>
                        </li>
                            <li class="nav-item">
                                <a href="https://github.com/stjude[cloud]/REPO_NAME/edit/master/docs/index.md" class="nav-link"><i class="fa fa-github"></i> Edit on GitHub</a>
                            </li>
                    </ul>
                </div>
            </div>
        </div>

        <div class="container">
            <div class="row">
                    <div class="col-md-3"><div class="navbar-light navbar-expand-md bs-sidebar hidden-print affix" role="complementary">
    <div class="navbar-header">
        <button type="button" class="navbar-toggler collapsed" data-toggle="collapse" data-target="#toc-collapse" title="Table of Contents">
            <span class="fa fa-angle-down"></span>
        </button>
    </div>

    
    <div id="toc-collapse" class="navbar-collapse collapse card bg-secondary">
        <ul class="nav flex-column">
            
            <li class="nav-item" data-level="1"><a href="#fuzzion2" class="nav-link">fuzzion2</a>
              <ul class="nav flex-column">
            <li class="nav-item" data-level="2"><a href="#build" class="nav-link">Build</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-level="2"><a href="#usage" class="nav-link">Usage</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-level="2"><a href="#patterns" class="nav-link">Patterns</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-level="2"><a href="#copyright" class="nav-link">COPYRIGHT</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-level="2"><a href="#license" class="nav-link">LICENSE</a>
              <ul class="nav flex-column">
              </ul>
            </li>
              </ul>
            </li>
        </ul>
    </div>
</div></div>
                    <div class="col-md-9" role="main">

<h1 id="fuzzion2">fuzzion2</h1>
<p>There are six programs in the Fuzzion2 suite: <code>fuzzion2</code>, <code>fuzzort</code>,
<code>fuzzum</code>, <code>fuzzion2html</code>, <code>fuzzall</code>, and <code>kmerank</code>.</p>
<p>The main program is <code>fuzzion2</code> which searches paired-end RNA or DNA to
identify read pairs that match fusion patterns.  The matching read pairs
are sorted by <code>fuzzort</code>; summarized by <code>fuzzum</code>; and converted to HTML by
<code>fuzzion2html</code> for viewing using a browser.  The <code>fuzzall</code> program
aggregates summaries produced by <code>fuzzum</code>.</p>
<p>The <code>fuzzion2</code> program requires as input a k-mer rank table.
Download the file named <code>fuzzion2_hg38_k15.krt</code> from
<a href="https://doi.org/10.5281/zenodo.6122447"><code>https://doi.org/10.5281/zenodo.6122447</code></a>.
It holds a 4-GB 15-mer rank table that was constructed from the GRCh38 human
reference genome.  Use this file only when searching human RNA or DNA.
The <code>kmerank</code> program is provided to construct k-mer rank tables for other
species.  Depending on the size of the table under construction, you may
need to run <code>kmerank</code> with as much as 21 GB of memory.</p>
<p><em>When <code>fuzzion2</code> reads a 4-GB k-mer rank table into memory, be sure to run
it with at least 5 GB of memory.</em>  Each of the other programs in the Fuzzion2
suite requires less than 1 GB of memory.</p>
<h2 id="build">Build</h2>
<pre><code>$ git clone https://github.com/stjude/fuzzion2.git
$ cd fuzzion2
$ make
</code></pre>
<p>This builds executable files for all six programs of the Fuzzion2 suite and
puts them in <code>build/bin</code>.</p>
<h4 id="dependencies">Dependencies</h4>
<p>These programs are written in C++ and compiled using <a href="https://gcc.gnu.org/">g++</a> version 6 or later.</p>
<p><a href="https://github.com/samtools/htslib">HTSlib</a> version 1.10.2 or later is used to read BAM files.
Set <code>CPATH</code> and <code>LIBRARY_PATH</code> before running <code>make</code> and set
<code>LD_LIBRARY_PATH</code> before running <code>fuzzion2</code>.</p>
<pre><code>$ HTSLIB=HTSlib-installation-directory
$ export CPATH=$CPATH:$HTSLIB/include
$ export LIBRARY_PATH=$LIBRARY_PATH:$HTSLIB/lib
$ export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$HTSLIB/lib
</code></pre>
<p><a href="https://www.gnu.org/software/gzip/">gunzip</a> is used to decompress gzipped FASTQ data.</p>
<h2 id="usage">Usage</h2>
<pre><code>Usage: fuzzion2 OPTION ... [filename ...] &gt; hits

These options are required:
  -pattern=filename   name of pattern input file
  -rank=filename      name of binary  input file containing the k-mer rank table

Specify -fastq1 and -fastq2, or -ifastq or -ubam, or list filenames on command line
  -fastq1=filename    name of FASTQ Read 1 input file
  -fastq2=filename    name of FASTQ Read 2 input file
  -ifastq=filename    name of interleaved FASTQ input file (may be /dev/stdin)
  -ubam=filename      name of unaligned Bam input file

The following are optional:
   N is a numeric value, e.g., -threads=4
  -maxins=N     maximum insert size in bases. . . . . . . . . . . . default 500
  -maxrank=N    maximum rank percentile of minimizers . . . . . . . default 99.9
  -maxtrim=N    maximum bases second read aligned ahead of first. . default 5
  -minbases=N   minimum percentile of matching bases. . . . . . . . default 90.0
  -minmins=N    minimum number of matching minimizers . . . . . . . default 1
  -minov=N      minimum overlap in number of bases. . . . . . . . . default 5
  -show=N       show best only (1) or all patterns (0) that match . default 1
  -single=N     show single-read (1) or just read-pair (0) matches. default 0
  -threads=N    number of threads . . . . . . . . . . . . . . . . . default 8
  -w=N          window length in number of bases. . . . . . . . . . default 10
</code></pre>
<p>The <code>-pattern</code> option is required and specifies the name of a text file containing
RNA or DNA patterns.  Look in the <code>patterns</code> directory for pattern files provided
with this distribution.</p>
<p>The <code>-rank</code> option is also required and specifies the name of a binary file
containing a k-mer rank table.  See the note above.</p>
<p><code>fuzzion2</code> expects read pairs as input; single-read formats are unsupported.
Each read pair is examined to see if it matches any of the patterns in the pattern
file.  Read pairs are obtained from:</p>
<ol>
<li>a pair of FASTQ files identified by the <code>-fastq1</code> and <code>-fastq2</code> options; or</li>
<li>an interleaved FASTQ file named by the <code>-ifastq</code> option; or</li>
<li>an unaligned Bam file named by the <code>-ubam</code> option; or</li>
<li>one or more files listed on the command line.</li>
</ol>
<p><code>fuzzion2</code> expects that the mates of a read pair are adjacent in interleaved FASTQ
and unaligned Bam files.  If a pair of FASTQ files is specified, the mates are
separated; "Read 1" mates are in one file and corresponding "Read 2" mates are
in another file.  If the name of a FASTQ file ends with ".gz", <code>fuzzion2</code> assumes
that the file is gzipped and uses <code>gunzip</code> to decompress it.</p>
<p>If <code>-fastq1</code>, <code>-fastq2</code>, <code>-ifastq</code>, and <code>-ubam</code> options are omitted, <code>fuzzion2</code>
looks for file names on the command line.  The named files can be any combination
of the above file types and can be listed in any order.  <code>fuzzion2</code> automatically
recognizes and pairs up corresponding "Read 1" and "Read 2" FASTQ files.</p>
<p>Each read pair that matches a pattern is a "hit" and is written along with the
pattern on three lines to the standard output stream.  In the following example,
<code>BCR-ABL1</code> is the name of the pattern.  The second column of the first line shows
the substring of the pattern sequence that matches the read pair.  The second and
third lines show the read name and entire sequence of each mate.</p>
<pre><code>pattern BCR-ABL1                 CATCCGTGGAGCTGCAGATGCTGACCAACTCGTGTGTGAAACTCCAGACTGTCCACAGCATTCCGCTGACCATCAATAA]GGAAGA[AGCCCTTCAGCGGCCAGTAGCATCTGACTTTGAGCCTCAGGGTCTGAGTGAAGCCGCTCGTTGGAACTCCAAGGAAAACCTTCTCGCTGGACCCAGTGAAAATGACCCCAACCTTTTCGTTGC
read EXAMPLE:1105:12909:66982/2  CATCCGTGGAGCTGCAGATGCTGACCAACTCGTGTGTGAAACTCCAGACTGTCCACAGCATTCCGCTGACCATCAATAAGGAAGAAGCCCTTCAGCGGCCA
read EXAMPLE:1105:12909:66982/1                                                                                                               TCTGACTTTGAGCCTCAGGGTCTGAGTGAAGCCGCTCGTTGGAACTCCAAGGAAAACCTTCTCGCTGGACCCAGTGAAAATGACCCCAACCTTTTCGTTGC
</code></pre>
<p>A final line is written to the standard output stream showing the total number
of read pairs processed by the program.</p>
<h4 id="additional-options">Additional Options</h4>
<p>You will normally not need to specify any of the options described below, but we mention
them here for completeness.</p>
<p>If there are similar patterns, a read pair may match more than one of them.  By default,
only the best match is reported.  Specify <code>-show=0</code> to see all of them.</p>
<p>By default, both mates of a read pair must match a pattern.  If there are pattern
sequences too short to match both mates, set <code>-single=1</code> to see also matches of one mate
to patterns.</p>
<p>If there are patterns consisting of very common k-mers, some matches may be missed.
In this case, the value of the <code>-maxrank</code> option should be increased.  By default, this
option is set to 99.9, which means the 0.1% most common k-mers in the reference genome are
ignored.  Increasing the value to 100.0 processes all k-mers, but the program may run
slowly.</p>
<p>The <code>-w</code> option specifies the window length; reducing this value increases the number
of minimizers representing each sequence.  The <code>-minmins</code> option specifies the minimum
number of minimizers shared by a read sequence and pattern sequence in a candidate match.</p>
<p>An alignment of a read pair to a pattern sequence is reported as a hit only if the
alignment has a reasonable insert size (not greater than the value of the <code>-maxins</code>
option); the read pair overlaps each side of the pattern by at least the value of the
<code>-minov</code> option; the percentage of bases in agreement on each side must be at least the
value of the <code>-minbases</code> option; and if the second read of the pair aligns ahead of
the first read, it is by no more than the value of the <code>-maxtrim</code> option.  Normally,
the first read will align ahead of the second read, but the latter option is provided
to accommodate imprecise adapter trimming.</p>
<h4 id="postprocessing">Postprocessing</h4>
<p>When multithreading is used (i.e., the value of the <code>-threads</code> option is greater
than 1), the order of the hits is indeterminate and a simple <code>diff</code> cannot be
used to compare output files.  It is therefore recommended to run the <code>fuzzort</code>
program to sort the hits.  This program sorts the hits by pattern name so that
the hits are in a determinate order (and <code>diff</code> can be used to compare files)
and the hits of a pattern are grouped together.</p>
<pre><code>Usage: fuzzort &lt; fuzzion2_hits &gt; sorted_hits
</code></pre>
<p>Run <code>fuzzion2html</code> to produce an HTML file that provides an attractive display
of hits when opened in a browser such as Google Chrome or Microsoft Edge.
SNPs, indels, and sequencing errors are highlighted in the display.</p>
<pre><code>Usage: fuzzion2html OPTION ... &lt; fuzzion2_hits &gt; html

The following are optional:
  -strong=N       minimum overlap of a strong match in #bases, default is 15
  -title=string   string to include in the title of the HTML page
</code></pre>
<p>The input to <code>fuzzort</code>, <code>fuzzion2html</code>, and <code>fuzzum</code> may be the output from a
single run of <code>fuzzion2</code> or the concatenation of outputs from multiple runs.</p>
<p><code>fuzzum</code> produces a tab-delimited summary of hits, and these summaries may be
aggregated using the <code>fuzzall</code> program.</p>
<pre><code>Usage: fuzzum OPTION ... &lt; fuzzion2_hits &gt; hit_summary

This option is required:
  -id=string   identifies the sample

The following is optional:
  -strong=N    minimum overlap of a strong match in #bases, default is 15

Usage: fuzzall OPTION fuzzum_filename ... &gt; pattern_summary

The following is optional:
  -dataset=name   name associated with this dataset
</code></pre>
<p>These summaries indicate the number of distinct read pairs matching each pattern,
and of those the number of "strong" versus "weak" matches.  A match is considered
to be strong if the alignment of the read pair to the pattern overlaps each side of
the pattern by at least N bases, where the value of N is given by the <code>-strong</code> option.
Otherwise, a match is regarded as weak due to insufficient overlap.  In <code>fuzzall</code>
output, each sample ID is followed by two numbers in parentheses, e.g., (24/22),
indicating the number of distinct matches (24) and strong matches (22).</p>
<h4 id="example-run">Example Run</h4>
<p>The <code>test</code> directory contains some files you can use to run a simple test:</p>
<pre><code>fuzzion2 -pattern=example_patterns.txt -rank=fuzzion2_hg38_k15.krt \
   -fastq1=example_input1.fq -fastq2=example_input2.fq &gt; my_output.txt

fuzzort &lt; my_output.txt &gt; my_sorted_output.txt

fuzzion2html -title=&quot;Fuzzion2 Example&quot; &lt; my_output.txt &gt; my_output.htm

fuzzum -id=example &lt; my_output.txt &gt; my_output_summary.txt
</code></pre>
<h2 id="patterns">Patterns</h2>
<p>Pattern files describe the nucleotide-level breakpoints of sequences of
interest, e.g., fusions, ITD (internal tandem duplication) boundaries,
or other targets.</p>
<h4 id="file-format">File Format</h4>
<p>A pattern file is formatted as tab-delimited text with a header line.  Two columns
must be present in the file:</p>
<ul>
<li>"pattern" contains the pattern identifier.  In our pattern set
  this is the gene pairing with a numbered suffix, e.g., BCR-ABL1-01 (note
  that these identifiers are not stable between releases).</li>
<li>"sequence" contains the sequence spanning the breakpoint.  A single
  pair of brackets is used in each sequence to indicate the boundaries
  of the breakpoint.  Two types of brackets may be
  used, square brackets ("]" and "[") for fusions, and curly brackets ("}" and "{")
  for ITD boundaries.  The right or closing bracket appears first, marking the end
  of the sequence upstream of the breakpoint (e.g., the first gene fusion partner),
  followed by the left or opening bracket indicating the start of the sequence
  downstream of the breakpoint (e.g., the second fusion gene partner). Sequence may
  optionally appear between brackets, indicating either interstitial sequence or a
  region of microhomology (i.e., a portion of the sequence that is ambiguous between
  the two sides of the breakpoint).  If curly brackets are used, <code>fuzzion2</code> will
  require at least one read to span the breakpoint.  Flanking sequence of 400-500 nt
  on either side of the breakpoint is recommended, or whatever length is appropriate
  for your sequencing's insert size.</li>
</ul>
<p>Additional columns may also be added to the pattern file for any other desired
information or annotations.  Below is an example pattern sequence for a BCR-ABL1
fusion:</p>
<pre><code>AGGGCGCCTTCCATGGAGACGCAGA][AGCCCTTCAGCGGCCAGTAGCATCT
</code></pre>
<p>This is a just a very short excerpt of the pattern sequence around the
breakpoint for illustrative purposes.  Square brackets appear in the
pattern, indicating a fusion event.  The sequence to the left of the
"]" is from BCR, the sequence to the right of the "[" is from ABL1.</p>
<h4 id="sources">Sources</h4>
<p>Pattern sets distributed with Fuzzion2 can be found in the "patterns"
subdirectory of this repository.  These sets were generated from fusion and ITD
data from various pediatric cancer projects and collaborations at St. Jude, such
as PCGP and NCI TARGET, as well as from clinical sequencing.  Patterns were also
generated from fusions described in the COSMIC database.  Pattern sets are works
in progress.</p>
<h4 id="programs-for-creating-pattern-files">Programs for Creating Pattern Files</h4>
<p>Coming soon.</p>
<h2 id="copyright">COPYRIGHT</h2>
<p>Copyright 2022 St. Jude Children's Research Hospital</p>
<h2 id="license">LICENSE</h2>
<p>Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at</p>
<p><a href="http://www.apache.org/licenses/LICENSE-2.0">http://www.apache.org/licenses/LICENSE-2.0</a></p>
<p>Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.</p></div>
            </div>
        </div>

        <footer class="col-md-12">
            <hr>
                <p>Â© 2021 St. Jude Children's Research Hospital</p>
            <p>Documentation built with <a href="https://www.mkdocs.org/">MkDocs</a>.</p>
        </footer>
        <script>
            var base_url = ".",
                shortcuts = {"help": 191, "next": 78, "previous": 80, "search": 83};
        </script>
        <script src="js/base.js" defer></script>
        <script src="search/main.js" defer></script>

        <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="searchModalLabel" aria-hidden="true">
    <div class="modal-dialog modal-lg">
        <div class="modal-content">
            <div class="modal-header">
                <h4 class="modal-title" id="searchModalLabel">Search</h4>
                <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
            </div>
            <div class="modal-body">
                <p>From here you can search these documents. Enter your search terms below.</p>
                <form>
                    <div class="form-group">
                        <input type="search" class="form-control" placeholder="Search..." id="mkdocs-search-query" title="Type search term here">
                    </div>
                </form>
                <div id="mkdocs-search-results" data-no-results-text="No results found"></div>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div><div class="modal" id="mkdocs_keyboard_modal" tabindex="-1" role="dialog" aria-labelledby="keyboardModalLabel" aria-hidden="true">
    <div class="modal-dialog">
        <div class="modal-content">
            <div class="modal-header">
                <h4 class="modal-title" id="keyboardModalLabel">Keyboard Shortcuts</h4>
                <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
            </div>
            <div class="modal-body">
              <table class="table">
                <thead>
                  <tr>
                    <th style="width: 20%;">Keys</th>
                    <th>Action</th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td class="help shortcut"><kbd>?</kbd></td>
                    <td>Open this help</td>
                  </tr>
                  <tr>
                    <td class="next shortcut"><kbd>n</kbd></td>
                    <td>Next page</td>
                  </tr>
                  <tr>
                    <td class="prev shortcut"><kbd>p</kbd></td>
                    <td>Previous page</td>
                  </tr>
                  <tr>
                    <td class="search shortcut"><kbd>s</kbd></td>
                    <td>Search</td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div>

    </body>
</html>

<!--
MkDocs version : 1.4.2
Build Date UTC : 2022-12-05 16:08:46.949112+00:00
-->
